File: C-Plus-Plus/cp/KMP_Algorithm.cpp
#include <iostream>
using namespace std;

int * getlps(char T[]){     //lps here holds longest prefix suffix
    int i;
  for(i=0;T[i]!='\0';i++);
  int len=i;                //Length
    int * lps= new int[len];
    lps[0]=0;
     i=1;   //index for 
    int j=0;
    while(i<len){
        if(T[i] == T[j]){     //where T is a pattern
            lps[i]=j+1;
            i++;
            j++;
        }else{      //mismatch after j matches
            if(j!=0){
                j=lps[j-1];
            }else{
                lps[i]=0;  //because j=0
                i++;
            }
        }
    }
    return lps;
}

int KMPSearch(char S[], char T[]) {    //Function to return final index where we found the substring 
   int i,j;
    for(i=0;S[i]!='\0';i++);
	int slen=i;            //Length of S
        for(j=0;T[j]!='\0';j++);
	int tlen=j;            //Length of T
     i=0;
     j=0;
    int * lps=getlps(T);
    
    while(i<slen&&j<tlen){
        if(S[i]==T[j]){     //means both are equal
            i++;
            j++;
        }else{         // (S[i] != T[j]) 
            if(j!=0){
                j=lps[j-1];   //do not increment i here
            }else{     //if (j==0)
                i++;
            }
        }
    }
    if(j==tlen){
        return i-j;
    }
    return -1;   //Pattern is not found
}



int main() {
    char S[1000], T[1000];
    cin.getline(S, 1000);
    cin.getline(T, 1000);
    cout << findString(S, T) << endl;          //Print the final answer
}

File: Java/Algorithms/KMPAlgorithm.java
package practice;

//JAVA program for implementation of KMP Algorithm

class KMP{ 
	void KMPSearch(String pat, String txt) 
	{ 
		int M = pat.length(); 
		int N = txt.length(); 

		int lps[] = new int[M]; 
		int j = 0; 
		computeLPSArray(pat, M, lps); 

		int i = 0; // index for txt[] 
		while (i < N) { 
			if (pat.charAt(j) == txt.charAt(i)) { 
				j++; 
				i++; 
			} 
			if (j == M) { 
				System.out.println("Found pattern "
								+ "at index " + (i - j)); 
				j = lps[j - 1]; 
			} 

			// mismatch after j matches 
			else if (i < N && pat.charAt(j) != txt.charAt(i)) { 
				if (j != 0) 
					j = lps[j - 1]; 
				else
					i = i + 1; 
			} 
		} 
	} 

	void computeLPSArray(String pat, int M, int lps[]) 
	{ 
		
		int len = 0; 
		int i = 1; 
		lps[0] = 0; 

		
		while (i < M) { 
			if (pat.charAt(i) == pat.charAt(len)) { 
				len++; 
				lps[i] = len; 
				i++; 
			} 
			else 
				if (len != 0) { 
					len = lps[len - 1]; 

				} 
				else 
				{ 
					lps[i] = len; 
					i++; 
				} 
			} 
		} 



	public static void main(String args[]) 
	{ 
		String txt = "ABABACDABABACDEFABABGHI"; 
		String pat = "ABABACDEFA"; 
		new KMP().KMPSearch(pat, txt); 
	} 
} 


